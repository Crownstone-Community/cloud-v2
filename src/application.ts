import {BootMixin} from '@loopback/boot';
import {ApplicationConfig} from '@loopback/core';
import {
  RestExplorerBindings,
  RestExplorerComponent,
} from '@loopback/rest-explorer';
import {RepositoryMixin} from '@loopback/repository';
import {RestApplication} from '@loopback/rest';
import {ServiceMixin} from '@loopback/service-proxy';
import path from 'path';
import {CrownstoneSequence} from './sequence';
import {
  AuthenticationComponent,
  registerAuthenticationStrategy
} from '@loopback/authentication';
import {
  AuthorizationComponent,
} from '@loopback/authorization';
import {UserService} from './services';
import {AccessTokenStrategy} from "./security/authentication-strategies/access-token-strategy";
import {SecurityTypes} from "./config";
import {SphereAuthorizationComponent} from "./security/authorization-strategies/authorization";

export interface PackageInfo {
  name: string;
  version: string;
  description: string;
}
const pkg: PackageInfo = require('../package.json');

export class CrownstoneCloud extends BootMixin(ServiceMixin(RepositoryMixin(RestApplication))) {
  constructor(options: ApplicationConfig = {}) {
    let executionPath = __dirname;
    if (options.customPath !== undefined) { executionPath = options.customPath; }
    let customPort = process.env.PORT || 3050;
    if (options.rest && options.rest.port !== undefined) {
      customPort = options.rest.port;
    }

    let customHost = process.env.HOST || '127.0.0.1';
    if (options.rest && options.rest.host !== undefined) {
      customHost = options.rest.host;
    }

    super({...options, rest: { ...options.rest, port: customPort, host: customHost }})

    this.api({
      openapi: '3.0.0',
      info: {title: pkg.name, version: pkg.version},
      paths: {},
      components: {securitySchemes: {
        [SecurityTypes.accessToken]: {
          type: 'apiKey',
          in: 'header',
          name:'Authorization'
        },
      }},
      servers:  [{url: '/api'}],
      security: [{[SecurityTypes.accessToken]: []}],
    });


    this.setUpBindings();
    // Bind authentication component related elements
    this.component(AuthenticationComponent);
    this.component(AuthorizationComponent);
    this.component(SphereAuthorizationComponent);


    // authentication
    registerAuthenticationStrategy(this, AccessTokenStrategy);

    // Set up the custom sequence
    this.sequence(CrownstoneSequence);

    // Set up default home page
    this.static('/', path.join(executionPath, '../public'));

    // Customize @loopback/rest-explorer configuration here
    this.configure(RestExplorerBindings.COMPONENT).to({ path: '/explorer' });
    this.component(RestExplorerComponent);

    this.projectRoot = executionPath;

    // We define both here to allow the testing framework to generate coverage over the ts files.
    this.bootOptions = {
      controllers: {
        dirs: ['controllers'],
        extensions: ['.controller.ts','.controller.js'],
        nested: true,
      },
      repositories: {
        dirs: ['repositories'],
        extensions: ['.repository.ts','.repository.js'],
        nested: true,
      },
      datasources: {
        dirs: ['datasources'],
        extensions: ['.datasource.ts','.datasource.js'],
        nested: true,
      },
      services: {
        dirs: ['services'],
        extensions: ['.service.ts','.service.js'],
        nested: true,
      },
    };
  }

  setUpBindings(): void {
    this.bind("UserService").toClass(UserService);
  }
}
