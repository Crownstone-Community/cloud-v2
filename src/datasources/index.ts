export * from './data.datasource';
export * from './users.datasource';
export * from './files.datasource';