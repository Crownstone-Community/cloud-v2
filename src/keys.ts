/**
 * This is used to get the name of an invoked endpoint in authorization.
 */

import {BindingKey} from '@loopback/core';

export const RESOURCE_ID = BindingKey.create<string>('resourceId');