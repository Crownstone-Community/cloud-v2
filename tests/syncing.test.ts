import {CrownstoneCloud} from "../src/application";
import {Client, createRestAppClient} from '@loopback/testlab';
import {clearTestDatabase, createApp, getRepositories} from "./helpers";
import {createHub, createSphere, createUser} from "./builders/createUserData";

let app    : CrownstoneCloud;
let client : Client;

beforeEach(async () => { await clearTestDatabase(); })
beforeAll(async () => {
  app    = await createApp()
  client = createRestAppClient(app);
})
afterAll(async () => { await app.stop(); })

test("AttemptSync", async () => {
  let user   = await createUser('test@test.com', 0);
  let sphere = await createSphere(user.id, 'mySphere', 0);
  let hub    = await createHub(sphere.id, 'myHub', 0);

  let repos = getRepositories();


  // await client.get("/users/sync").expect(401)
});