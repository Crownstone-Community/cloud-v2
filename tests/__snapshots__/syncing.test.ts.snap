// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync FULL 1`] = `
Object {
  "spheres": Object {
    "dbId:SphereRepository:1": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereRepository:1",
          "name": "mySphere",
          "uid": 209,
          "updatedAt": "1970-01-01T00:00:00.000Z",
          "uuid": "ibeaconUUID2000",
        },
        "status": "VIEW",
      },
      "hubs": Object {
        "dbId:HubRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:HubRepository:1",
              "name": "myHub",
              "sphereId": "dbId:SphereRepository:1",
              "token": "helloI'mAHub",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
      "locations": Object {
        "dbId:LocationRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:LocationRepository:1",
              "name": "location",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
      "stones": Object {
        "dbId:StoneRepository:1": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:1": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:1",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:1",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:1",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2009",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:1",
              "location": Object {
                "id": "dbId:LocationRepository:1",
                "name": "location",
              },
              "locationId": "dbId:LocationRepository:1",
              "major": 2010,
              "minor": 2011,
              "name": "stone1",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:StoneRepository:2": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:2": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:2",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:2",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:2",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2014",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:2",
              "major": 2015,
              "minor": 2016,
              "name": "stone2",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:StoneRepository:3": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:3": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:3",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:3",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:3",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2019",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:3",
              "major": 2020,
              "minor": 2021,
              "name": "stone3",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
    },
  },
  "user": Object {
    "data": Object {
      "accountCreationPending": false,
      "createdAt": "1970-01-01T00:00:00.000Z",
      "email": "test@test.com",
      "emailVerified": false,
      "id": "dbId:UserRepository:1",
      "language": "en_us",
      "new": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
      "uploadDeviceDetails": true,
      "uploadLocation": true,
      "uploadSwitchState": true,
    },
    "status": "VIEW",
  },
}
`;

exports[`Sync REQUEST with empty body 1`] = `
Object {
  "spheres": Object {
    "dbId:SphereRepository:1": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereRepository:1",
          "name": "mySphere",
          "uid": 209,
          "updatedAt": "1970-01-01T00:00:00.000Z",
          "uuid": "ibeaconUUID2000",
        },
        "status": "NEW_DATA_AVAILABLE",
      },
      "hubs": Object {
        "dbId:HubRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:HubRepository:1",
              "name": "myHub",
              "sphereId": "dbId:SphereRepository:1",
              "token": "helloI'mAHub",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
      "locations": Object {
        "dbId:LocationRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:LocationRepository:1",
              "name": "location",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
      "stones": Object {
        "dbId:StoneRepository:1": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:1": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:1",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:1",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:1",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "NEW_DATA_AVAILABLE",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2009",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:1",
              "location": Object {
                "id": "dbId:LocationRepository:1",
                "name": "location",
              },
              "locationId": "dbId:LocationRepository:1",
              "major": 2010,
              "minor": 2011,
              "name": "stone1",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
        "dbId:StoneRepository:2": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:2": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:2",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:2",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:2",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "NEW_DATA_AVAILABLE",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2014",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:2",
              "major": 2015,
              "minor": 2016,
              "name": "stone2",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
        "dbId:StoneRepository:3": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:3": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:3",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:3",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:3",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "NEW_DATA_AVAILABLE",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2019",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:3",
              "major": 2020,
              "minor": 2021,
              "name": "stone3",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
    },
  },
  "user": Object {
    "data": Object {
      "accountCreationPending": false,
      "createdAt": "1970-01-01T00:00:00.000Z",
      "email": "test@test.com",
      "emailVerified": false,
      "id": "dbId:UserRepository:1",
      "language": "en_us",
      "new": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
      "uploadDeviceDetails": true,
      "uploadLocation": true,
      "uploadSwitchState": true,
    },
    "status": "NEW_DATA_AVAILABLE",
  },
}
`;

exports[`Sync REQUEST with empty body and scope 1`] = `
Object {
  "spheres": Object {
    "dbId:SphereRepository:1": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereRepository:1",
          "name": "mySphere",
          "uid": 209,
          "updatedAt": "1970-01-01T00:00:00.000Z",
          "uuid": "ibeaconUUID2000",
        },
        "status": "NEW_DATA_AVAILABLE",
      },
      "hubs": Object {
        "dbId:HubRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:HubRepository:1",
              "name": "myHub",
              "sphereId": "dbId:SphereRepository:1",
              "token": "helloI'mAHub",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
    },
  },
}
`;

exports[`Sync REQUEST with request body 1`] = `
Object {
  "spheres": Object {
    "dbId:SphereRepository:1": Object {
      "data": Object {
        "status": "IN_SYNC",
      },
      "features": Object {},
      "hubs": Object {
        "dbId:HubRepository:1": Object {
          "data": Object {
            "status": "IN_SYNC",
          },
        },
      },
      "locations": Object {
        "dbId:LocationRepository:1": Object {
          "data": Object {
            "status": "IN_SYNC",
          },
        },
      },
      "scenes": Object {
        "my-new-scene-id": Object {
          "data": Object {
            "status": "DELETED",
          },
        },
      },
      "stones": Object {
        "dbId:StoneRepository:1": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:1": Object {
              "data": Object {
                "status": "IN_SYNC",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:1": Object {
                  "data": Object {
                    "status": "IN_SYNC",
                  },
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:1": Object {
              "data": Object {
                "status": "IN_SYNC",
              },
            },
          },
          "data": Object {
            "status": "IN_SYNC",
          },
        },
        "dbId:StoneRepository:2": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:2": Object {
                  "data": Object {
                    "data": Object {
                      "abilityId": "dbId:StoneAbilityRepository:2",
                      "createdAt": "1970-01-01T00:00:00.000Z",
                      "id": "dbId:StoneAbilityPropertyRepository:2",
                      "sphereId": "dbId:SphereRepository:1",
                      "stoneId": "dbId:StoneRepository:2",
                      "type": "smoothDimming",
                      "updatedAt": "1970-01-01T00:00:00.000Z",
                      "value": "true",
                    },
                    "status": "NEW_DATA_AVAILABLE",
                  },
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
            },
          },
          "data": Object {
            "status": "IN_SYNC",
          },
        },
        "dbId:StoneRepository:3": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:3": Object {
                  "data": Object {
                    "data": Object {
                      "abilityId": "dbId:StoneAbilityRepository:3",
                      "createdAt": "1970-01-01T00:00:00.000Z",
                      "id": "dbId:StoneAbilityPropertyRepository:3",
                      "sphereId": "dbId:SphereRepository:1",
                      "stoneId": "dbId:StoneRepository:3",
                      "type": "smoothDimming",
                      "updatedAt": "1970-01-01T00:00:00.000Z",
                      "value": "true",
                    },
                    "status": "NEW_DATA_AVAILABLE",
                  },
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "NEW_DATA_AVAILABLE",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2019",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:3",
              "major": 2020,
              "minor": 2021,
              "name": "stone3",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
      "toons": Object {},
      "trackingNumbers": Object {},
    },
  },
  "user": Object {
    "data": Object {
      "accountCreationPending": false,
      "createdAt": "1970-01-01T00:00:00.000Z",
      "email": "test@test.com",
      "emailVerified": false,
      "id": "dbId:UserRepository:1",
      "language": "en_us",
      "new": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
      "uploadDeviceDetails": true,
      "uploadLocation": true,
      "uploadSwitchState": true,
    },
    "status": "NEW_DATA_AVAILABLE",
  },
}
`;

exports[`Sync REQUEST with request body and new items created in the cloud 1`] = `
Object {
  "abilities": Object {
    "dbId:StoneAbilityRepository:1": Object {
      "data": Object {
        "status": "IN_SYNC",
      },
      "properties": Object {
        "dbId:StoneAbilityPropertyRepository:1": Object {
          "data": Object {
            "status": "IN_SYNC",
          },
        },
        "dbId:StoneAbilityPropertyRepository:4": Object {
          "data": Object {
            "data": Object {
              "abilityId": "dbId:StoneAbilityRepository:1",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneAbilityPropertyRepository:4",
              "sphereId": "dbId:SphereRepository:1",
              "stoneId": "dbId:StoneRepository:1",
              "type": "test",
              "updatedAt": "1970-01-01T00:00:00.000Z",
              "value": "hello",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
    },
  },
  "behaviours": Object {
    "dbId:StoneBehaviourRepository:1": Object {
      "data": Object {
        "status": "IN_SYNC",
      },
    },
  },
  "data": Object {
    "status": "IN_SYNC",
  },
}
`;

exports[`Sync REQUEST with request body and new items created in the cloud 2`] = `
Object {
  "abilities": Object {
    "dbId:StoneAbilityRepository:1": Object {
      "data": Object {
        "status": "IN_SYNC",
      },
      "properties": Object {
        "dbId:StoneAbilityPropertyRepository:1": Object {
          "data": Object {
            "status": "IN_SYNC",
          },
        },
        "dbId:StoneAbilityPropertyRepository:4": Object {
          "data": Object {
            "data": Object {
              "abilityId": "dbId:StoneAbilityRepository:1",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneAbilityPropertyRepository:4",
              "sphereId": "dbId:SphereRepository:1",
              "stoneId": "dbId:StoneRepository:1",
              "type": "test",
              "updatedAt": "1970-01-01T00:00:00.000Z",
              "value": "hello",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
    },
    "dbId:StoneAbilityRepository:4": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "enabled": true,
          "id": "dbId:StoneAbilityRepository:4",
          "sphereId": "dbId:SphereRepository:1",
          "stoneId": "dbId:StoneRepository:1",
          "syncedToCrownstone": true,
          "type": "test",
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        "status": "NEW_DATA_AVAILABLE",
      },
      "properties": Object {
        "dbId:StoneAbilityPropertyRepository:5": Object {
          "data": Object {
            "data": Object {
              "abilityId": "dbId:StoneAbilityRepository:4",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneAbilityPropertyRepository:5",
              "sphereId": "dbId:SphereRepository:1",
              "stoneId": "dbId:StoneRepository:1",
              "type": "test",
              "updatedAt": "1970-01-01T00:00:00.000Z",
              "value": "hello",
            },
            "status": "NEW_DATA_AVAILABLE",
          },
        },
      },
    },
  },
  "behaviours": Object {
    "dbId:StoneBehaviourRepository:1": Object {
      "data": Object {
        "status": "IN_SYNC",
      },
    },
  },
  "data": Object {
    "status": "IN_SYNC",
  },
}
`;

exports[`Sync REQUEST with request body and new items from app, propagate new 1`] = `
Object {
  "abilities": Object {
    "who": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "enabled": true,
          "id": "dbId:StoneAbilityRepository:4",
          "sphereId": "dbId:SphereRepository:1",
          "stoneId": "dbId:StoneRepository:4",
          "syncedToCrownstone": true,
          "type": "test",
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        "status": "CREATED_IN_CLOUD",
      },
      "properties": Object {
        "ack": Object {
          "data": Object {
            "data": Object {
              "abilityId": "dbId:StoneAbilityRepository:4",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneAbilityPropertyRepository:4",
              "sphereId": "dbId:SphereRepository:1",
              "stoneId": "dbId:StoneRepository:4",
              "type": "proppy",
              "updatedAt": "1970-01-01T00:00:00.000Z",
              "value": "Yes",
            },
            "status": "CREATED_IN_CLOUD",
          },
        },
      },
    },
  },
  "behaviours": Object {},
  "data": Object {
    "data": Object {
      "address": "yes!",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "id": "dbId:StoneRepository:4",
      "major": 2024,
      "minor": 2025,
      "sphereId": "dbId:SphereRepository:1",
      "updatedAt": "1970-01-01T00:00:00.000Z",
    },
    "status": "CREATED_IN_CLOUD",
  },
}
`;
